### PR TITLE
yul: Add new optimizer rules for create and create2 builtins

### LINKED ISSUE #7356: [yul] optimizer assertion inside SimplificationRules pattern matcher fails
## Description

```
{
 for {} and(create(1,0,0), msize()) {} 
 {
 }
}
```

fails this assertion

https://github.com/ethereum/solidity/blob/049e9500858324472d1bf38d1803bf3625fb2dec/libyul/optimiser/SimplificationRules.cpp#L160

## Environment

- Compiler version: latest develop
- Target EVM version (as per compiler settings): petersburg

## Steps to Reproduce

```
$ solc --strict-assembly --optimize <code.yul>
Exception in assembler: /home/bhargava/work/github/solidity/libyul/optimiser/SimplificationRules.cpp(160): Throw in function bool yul::Pattern::matches(const yul::Expression &, const yul::Dialect &, const map<yul::YulString, const yul::Expression *> &) const
Dynamic exception type: boost::exception_detail::clone_impl<yul::OptimizerException>
std::exception::what:
[dev::tag_comment*] =
```

### PR DESCRIPTION
**Pull request by bshastry**
fixes #7356

### CODE REVIEW

File: libevmasm/RuleList.h [hunk 1]
```diff
@@ -565,28 +564,87 @@ std::vector<SimplificationRule<Pattern>> simplificationRuleListPart8(
...
+	Pattern W,
+	Pattern X,
+	Pattern Y,
+	Pattern Z
+)
+{
+	std::vector<SimplificationRule<Pattern>> rules;
+
+	u256 const mask = (u256(1) << 160) - 1;
+	rules.push_back({
```
Comment (chriseth) [inline libevmasm/RuleList.h:593]: You could still put that in the same function (i.e. `Part9`)

**Comment by chriseth** [comment]
Looks good apart from the one comment!

**Comment by bshastry** [comment]
Added yul optimizer tests for the newly created rules. 

@chriseth Should I add legacy optimizer tests as well here? 

https://github.com/ethereum/solidity/blob/develop/test/libevmasm/Optimiser.cpp

**Comment by chriseth** [comment]
If that is not too much work, you can try a legacy optimizer test that uses `create` but takes its argument from the stack after a jumpdest, so that the argument is not directly visible.

**Comment by bshastry** [comment]
> jumpdest

Currently investigating how I can get past the following failure

```
test/libevmasm/Optimiser.cpp(74): fatal error: in "Optimiser/cse_create_mask": critical check cse.feedItems(input.begin(), input.end(), usesMsize) == input.end() has failed
```

caused most likely by the second condition here https://github.com/ethereum/solidity/blob/96b6b4565801fc52997ea212edcb695eefc36315/libevmasm/SemanticInformation.cpp#L72

because `create` contains `> 2` arguments

**Review by chriseth — APPROVED** [review]

**Comment by bshastry** [comment]
The create2 yul optimizer test was failing because the builtin is only available `>= constantinople`. This is now fixed.

**Pull request merged**